_posts/2019-06-05-quarkus-and-web-ui-development-mode.adoc

```diff
@@ -0,0 +1,110 @@
+
+==== Adjustments Summary
+
+The example project has already been set up with the needed adjustments to do this. These are the main things which need doing:
+
+. Make sure we handle the Angular routes in the UI rather than on the back-end. This is important for the bundled version (running on port 8080).
+. Enable the Angular proxy
+.. Signal to the back-end that the front-end is running in a proxy, i.e. not served by us. This is important for when the back-end needs to redirect to a page in the front-end.
+
+We will not go into details of how the application is set up, but will instead focus on the changes needed to make this work well.
+
+==== Handle Angular Routes
+We need to add a filter, https://github.com/kabir/blog-quarkus-ui-development/blob/master/src/main/java/org/kabir/quarkus/ui/AngularRouteFilter.java[AngularRouteFilter]. The essence of its `doFilter()` method is:
+[source,java]
+----
+chain.doFilter(request, response);
+
+if (response.getStatus() == 404) {
+    String path = request.getRequestURI().substring(
+            request.getContextPath().length()).replaceAll("[/]+$", "");
+    if (!FILE_NAME_PATTERN.matcher(path).matches()) {
+        // We could not find the resource, i.e. it is not anything known to the server (i.e. it is not a REST
+        // endpoint or a servlet), and does not look like a file so try handling it in the front-end routes.
+        request.getRequestDispatcher("/").forward(request, response);
+    }
+}
+
+----
+Without this filter, if you try to go to one of the Angular routes directly (e.g if you try to refresh when on http://localhost:8080/rest which is the `Rest` page we saw earlier) Quarkus will think it is a file, a REST endpoint or a servlet. Since it has no knowledge of anything called `/rest` you end up with a 404 (Not Found) status. `AngularRouteFilter` checks the status code, and if it is 404 and does not seem to be a file, we attempt to handle it in Angular.
+
+
+==== Set up the Angular proxy
+We configure the proxy in https://github.com/kabir/blog-quarkus-ui-development/blob/master/webapp/proxy.conf.json[proxy.conf.json]. All REST calls to anything under `/api` will be passed to the back-end running on port 8080. To run the Angular development server with this configuration, we have added a `proxy` configuration to the `scripts` section of https://github.com/kabir/blog-quarkus-ui-development/blob/master/webapp/package.json[`package.json`].
+
+In our case, we have a servlet which needs to redirect back to the front-end (something I found I needed to implement OAuth in my main project). That has a check for the `-Dui.proxy` system property we saw earlier when handling the `/callback` path in https://github.com/kabir/blog-quarkus-ui-development/blob/master/src/main/java/org/kabir/quarkus/ui/SampleServlet.java[`SampleServlet`]. If this property is set, we prepend `https://localhost:4200` (the address of the Angular proxy) to the redirect URL if we find the proxy is running on port 4200.
```

Comment:
Any reason you don't read the ui.* properties via the `@Inject @ConfigProperty`? That way you coudl read them via applciation.properties, system properties and enviornment properties equally

Reply:
Also I did not understnand why this servlet was only answering to `/servlet/` and not everything. As you can see I'm still a bit lost on the app flow exchange.

Reply:
Not really, just to keep it simple (as well as being slightly ignorant of MP config). We only ever want this to be triggered via development mode though, it should never be used in production.